x/msgfee/model.go

```diff
@@ -0,0 +1,75 @@
+package msgfee
+
+import (
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/coin"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/orm"
+)
+
+var _ orm.CloneableData = (*MsgFee)(nil)
+
+func (mf *MsgFee) Validate() error {
+	if mf.MsgPath == "" {
+		return errors.Wrap(errors.ErrInvalidModel, "invalid message path")
+	}
+	if mf.Fee.IsZero() {
+		return errors.Wrap(errors.ErrInvalidModel, "invalid fee")
+	}
+	if err := mf.Fee.Validate(); err != nil {
+		return errors.Wrap(err, "invalid fee")
+	}
+	return nil
+}
+
+func (mf *MsgFee) Copy() orm.CloneableData {
+	return &MsgFee{
+		MsgPath: mf.MsgPath,
+		Fee:     *mf.Fee.Clone(),
+	}
+}
+
+type MsgFeeBucket struct {
+	orm.Bucket
+}
+
+// NewMsgFeeBucket returns a bucket for keeping track of fees for eeach message
+// type. Message fees are indexed by the corresponding message path.
+func NewMsgFeeBucket() *MsgFeeBucket {
+	b := orm.NewBucket("msgfee", orm.NewSimpleObj(nil, &MsgFee{}))
+	return &MsgFeeBucket{
+		Bucket: b,
+	}
+}
+
+// Create adds given message fee instance to the store.
+func (b *MsgFeeBucket) Create(db weave.KVStore, mf *MsgFee) (orm.Object, error) {
+	key := []byte(mf.MsgPath)
+	obj := orm.NewSimpleObj(key, mf)
+	return obj, b.Bucket.Save(db, obj)
+}
+
+// Save persists the state of a given revenue entity.
+func (b *MsgFeeBucket) Save(db weave.KVStore, obj orm.Object) error {
```

Comment:
Note if I change the path and then save again, the key and the value will be different...

Maybe ensure that `mf.MsgPath == string(obj.Key())` for safety